tests/poseidon2/testIo.nim

```diff
@@ -0,0 +1,36 @@
+import std/unittest
+import std/sequtils
+import constantine/math/io/io_bigints
+import constantine/math/arithmetic
+import poseidon2/types
+import poseidon2/io
+
+suite "unmarshalling":
+
+  const F123456 = F.fromBig(B.fromHex("0x123456", bigEndian))
+
+  test "converts big endian bytes into field elements":
+    let bytes = [0x12'u8, 0x34'u8, 0x56'u8]
+    var padded: array[31, byte]
+    padded[^3..^1] = bytes
```

Comment:
Ok so what I really meant, is that I find the convention of padding the 3-byte sequence `[a,b,c]` to the 31-byte sequence `[0,0,0...,0,a,b,c]` really weird. We normally don't pad byte sequences _on the left_. If you pad instead to the more natural `[a,b,c,0,0...,0]` then the original code looks wrong to me in the big-endian case. 

In any case I suggest to:
- do the padding for last 31 byte chunk explicitly, instead letting constantine doing, which is implicit in our code
- document the padding convention
- write hash tests concrete against not-multiple-of-31 byte sequences for both the big-endian and little-endian case
- optional, put quite preferable: figure out a padding convention which does not allow hashes of different length byte sequences to coincide (except with negligible probability). Example: use the length measured in bytes as either the first or last field element). Note that the traditional hashes like SHA256 employ padding techniques to have to property.

Similar problems appear with the Merkle root, but there we also have to coordinate with the rest of codebase to be compatible.

I can provide a self-contained pure C reference implementation if there is need for it.

Reply:
Ok, sounds like a plan :+1:  I have a few remarks:

> * write hash tests concrete against not-multiple-of-31 byte sequences for both the big-endian and little-endian case

We should do tests for big-endian and little-endian for the `unmarshal()` calls, but not for the `merkleRoot()` hashes. We currently do not allow for calculation of hashes with big-endian encoding, and I don't see a reason why we should support it?

> * optional, put quite preferable: figure out a padding convention which does not allow hashes of different length byte sequences to coincide (except with negligible probability). Example: use the length measured in bytes as either the first or last field element). Note that the traditional hashes like SHA256 employ padding techniques to have this property.

I'd like to take a stab at that, but in a different PR.

> I can provide a self-contained pure C reference implementation if there is need for it.

I'd rather have it in Nim :stuck_out_tongue_winking_eye:

Reply:
>  but not for the merkleRoot() hashes

I somewhat disagree in the sense that I think after you added the choice of endianness, we should do all of it. If you haven't added it at all, I wouldn't ask for it :) But yeah maybe it's enough to test `unmarshal` for now.

> I'd like to take a stab at that, but in a different PR.

But first let's discuss and fix the conventions. It should be the same across the the codebase.

> I'd rather have it in Nim 

The point is exactly _not having it in Nim_, so that it's a real check and not a self-consistency check :)

But we can get away with some test vectors and a link to a different implementation which generated those test vectors.

Reply:
Ok, I added the right-padding of big-endian bytes in 76d08b8, but it looked really weird. We'd have to pad with 0's to the right to get 31 bytes, and then add an additional 0 byte to the left to get to 32 bytes. This is because the most significant byte needs to be 0, otherwise we might get numbers outside of the field.

In the end I decided to drop the big-endian encoding; we don't really need it, and it makes everything more complex.

Reply:
This is indeed weird. This shouldn't be the case just from reading the code, but I agree that we don't need it 

(could be at some point possibly useful if we want to interface with any third party software which decided to use big-endian, but such a thing existing seems unlikely; I think on all the wide-spread commodity hardware little-endian is the natural choice)

Maybe I will look into it when I have a working environment set up, just to clarify what happens.